working/control-flow-collections/feature-specification.md

```diff
@@ -0,0 +1,1021 @@
+    IconButton(icon: Icon(Icons.menu))
+  ]);
+}
+```
+
+Now, the fact that Row's `children` parameter has type `List<Widget>` causes
+that to be the inferred type of the list. We're able to do this because the
+entire list creation is now a single expression so it can be moved right into
+the constructor call for `Row()`.
+
+## Syntax
+
+We extend the list grammar to allow *control flow elements* in addition to
+regular elements:
+
+```
+listLiteral:
+  const? typeArguments? '[' listElementList? ']'
+  ;
+
+listElementList:
+  listElement ( ',' listElement )* ','?
+  ;
+
+listElement:
+  expression |
+  'if' '(' expression ')' listElement ( 'else' listElement )? |
+  'await'? 'for' '(' forLoopParts ')' listElement
+  ;
+```
+
+Instead of `expressionList`, this uses a new `listElementList` rule since
+`expressionList` is used elsewhere in the grammar like argument lists where
+control flow isn't allowed.
+
+Each element in a list can be one of a few things:
+
+* A normal expression.
+* An `if` element.
+* A `for` element.
+
+The body of `if` and `for` elements use `listElement`, not `expression`, which
+allows nesting.
+
+The changes for map literals are similar:
+
+```
+mapLiteral:
+  const? typeArguments? '{' mapLiteralEntryList? '}' ;
+
+mapLiteralEntryList:
+  mapLiteralEntry ( ',' mapLiteralEntry )* ','?
+  ;
+
+mapLiteralEntry:
+  expression ':' expression |
+  'if' '(' expression ')' mapLiteralEntry ( 'else' mapLiteralEntry )? |
+  'await'? 'for' '(' forLoopParts ')' mapLiteralEntry
+  ;
+```
+
+## Static Semantics
+
+Let the *element type* of a list literal be the static type of the type argument
+used to create the list. So `<int>[]` has an element type of `int`. It may be
+explicit or filled in by type inference. So `[1, 2.0]` has an element type of
+`num`.
+
+Let the *key type* and *value type* of a map literal be the corresponding
+static types of the type arguments for a map literal. So `<int, String>{}` and
+`{1: "s"}` both have a key type of `int` and a value type of `String`.
+
+Let the *body elements* of an `if` element be the "then" element and the "else"
+element if there is one. Let the *body elements* of a `for` element be the
+single element it contains.
+
+### Scoping
+
+Both styles of `for` element may introduce a local variable, as in:
+
+```dart
+[
+  for (var i = 1; i < 4; i++) i,
+  for (var i in [1, 2, 3]) i
+]
+```
+
+If a `for` element declares a variable, then a new namespace is created where
+that variable is defined. The body of the `for` element is resolved and
+evaluated in that namespace. The variable goes out of scope at the end of the
+element's body.
+
+Each iteration of the loop binds a new fresh variable:
+
+```dart
+var closures = [for (var i = 1; i < 4; i++) () => i];
+for (var closure in closures) print(closure());
+// Prints "1", "2", "3".
+```
+
+### Static errors
+
+The static semantics of collection `if` and `for` mostly follow their statement
+analogues.
```

Comment:
Is it worth spelling out that it is a static error to have a list element inside of a map literal, or a map element inside of a list literal, or is this considered a parse error?